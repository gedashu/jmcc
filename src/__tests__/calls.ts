import JMCC from '../core';

const jmcc = new JMCC();

describe('вызовы', () => {
  describe('функций', () => {
    describe('встроенных объектов', () => {
      test('обычных', () => {
        const source = `
          var a = variable::create_list([1, 2])
          var c = variable::create_list(values = [1, 2])
          
          player::message("hello!", merging = "CONCATENATION")
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('с аргументом массивом', () => {
        const source = `
          player::message(["hello", "world!"], merging = "SPACES")
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('с неизвестным аргументом', () => {
        const source = `
          var a = variable::create_list(b = 1)
        `;

        expect(() => jmcc.compileToJson(source)).toThrowErrorMatchingSnapshot();
      });

      test('с неподходящим типом аргумента', () => {
        const source = `
          var a = variable::clamp(0, 1, "2")
        `;

        expect(() => jmcc.compileToJson(source)).toThrowErrorMatchingSnapshot();
      });

      test('проверяющих значение', () => {
        const source = `
          if (player::is_holding(item("stone"))) {
            player::message("hello")
          }
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('повторений', () => {
        const source = `
          repeat::forever() {
            player::message("hello")
          }
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('повторений с условием', () => {
        const source = `
          repeat::while(player::is_holding(item("stone"))) {
            player::message("hello")
          }
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('повторений с лямбдами', () => {
        const source = `
          repeat::on_range(0, 2) { i ->
            player::message("idx: $i")
          }
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });
    });

    describe('переменных', () => {
      test('обычных', () => {
        const source = `
          var a = 5
          a.increment(1)
          a.decrement(2)
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('с установкой значения', () => {
        const source = `
          var a = 5
          var b = a.clamp(1, 2)
          var c = a.clamp(min = 1, max = 2)
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('с неизвестным аргументом', () => {
        const source = `
          var a = 5
          var b = a.clamp(min = 1, min2 = 2)
        `;

        expect(() => jmcc.compileToJson(source)).toThrowErrorMatchingSnapshot();
      });

      test('с неподходящим типом аргумента', () => {
        const source = `
          var a = 5
          var b = a.clamp(1, "2")
        `;

        expect(() => jmcc.compileToJson(source)).toThrowErrorMatchingSnapshot();
      });
    });

    describe('фабрик', () => {
      test('обычных', () => {
        const source = `
          var a = location(1, 2, 3)
          var b = location(1, 2, 3, 4, 5)
        `;

        expect(jmcc.compileToJson(source)).toMatchSnapshot();
      });

      test('с неподходящим типом аргумента', () => {
        const source = `
          var a = location(1, 2, "3")
        `;

        expect(() => jmcc.compileToJson(source)).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('массивов', () => {
    test('обычных', () => {
      const source = `
        var a = [1, 2]
        var b = a[0]
      `;

      expect(jmcc.compileToJson(source)).toMatchSnapshot();
    });

    test('с установкой', () => {
      const source = `
        var a = [1, 2]
        a[1] = 2
      `;

      expect(jmcc.compileToJson(source)).toMatchSnapshot();
    });
  });
});
