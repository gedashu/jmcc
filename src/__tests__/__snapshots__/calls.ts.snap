// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`вызовы массивов обычных 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_create_list",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "values",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "number": 1,
                    "type": "number",
                  },
                  Object {
                    "number": 2,
                    "type": "number",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
          ],
        },
        Object {
          "action": "set_variable_get_list_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "list",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 0,
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы массивов с установкой 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_create_list",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "values",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "number": 1,
                    "type": "number",
                  },
                  Object {
                    "number": 2,
                    "type": "number",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
          ],
        },
        Object {
          "action": "set_variable_set_list_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "list",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов обычных 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_create_list",
          "values": Array [
            Object {
              "name": "values",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "number": 1,
                    "type": "number",
                  },
                  Object {
                    "number": 2,
                    "type": "number",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_create_list",
          "values": Array [
            Object {
              "name": "values",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "number": 1,
                    "type": "number",
                  },
                  Object {
                    "number": 2,
                    "type": "number",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "c",
              },
            },
          ],
        },
        Object {
          "action": "player_send_message",
          "values": Array [
            Object {
              "name": "messages",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "text": "hello!",
                    "type": "text",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "merging",
              "value": Object {
                "enum": "CONCATENATION",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов повторений 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "repeat_forever",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "hello",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов повторений с лямбдами 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "repeat_on_range",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "idx: %var_local(i)",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "start",
              "value": Object {
                "number": 0,
                "type": "number",
              },
            },
            Object {
              "name": "end",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "i",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов повторений с условием 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "repeat_while",
          "conditional": Object {
            "action": "if_player_is_holding",
            "is_inverted": false,
            "operations": Array [],
            "values": Array [
              Object {
                "name": "items",
                "value": Object {
                  "type": "array",
                  "values": Array [
                    Object {
                      "item": "{\\"id\\":\\"stone\\",\\"Count\\":1}",
                      "type": "item",
                    },
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                    Object {},
                  ],
                },
              },
              Object {
                "name": "hand_slot",
                "value": Object {
                  "enum": "EITHER_HAND",
                  "type": "enum",
                },
              },
              Object {
                "name": "comparison_mode",
                "value": Object {
                  "enum": "EXACTLY",
                  "type": "enum",
                },
              },
            ],
          },
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "hello",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "items",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "item": "{\\"id\\":\\"stone\\",\\"Count\\":1}",
                    "type": "item",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "hand_slot",
              "value": Object {
                "enum": "EITHER_HAND",
                "type": "enum",
              },
            },
            Object {
              "name": "comparison_mode",
              "value": Object {
                "enum": "EXACTLY",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов проверяющих значение 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "if_player_is_holding",
          "is_inverted": false,
          "operations": Array [
            Object {
              "action": "player_send_message",
              "values": Array [
                Object {
                  "name": "messages",
                  "value": Object {
                    "type": "array",
                    "values": Array [
                      Object {
                        "text": "hello",
                        "type": "text",
                      },
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                      Object {},
                    ],
                  },
                },
                Object {
                  "name": "merging",
                  "value": Object {
                    "enum": "SPACES",
                    "type": "enum",
                  },
                },
              ],
            },
          ],
          "values": Array [
            Object {
              "name": "items",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "item": "{\\"id\\":\\"stone\\",\\"Count\\":1}",
                    "type": "item",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "hand_slot",
              "value": Object {
                "enum": "EITHER_HAND",
                "type": "enum",
              },
            },
            Object {
              "name": "comparison_mode",
              "value": Object {
                "enum": "EXACTLY",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов с аргументом массивом 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "player_send_message",
          "values": Array [
            Object {
              "name": "messages",
              "value": Object {
                "type": "array",
                "values": Array [
                  Object {
                    "text": "hello",
                    "type": "text",
                  },
                  Object {
                    "text": "world!",
                    "type": "text",
                  },
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                  Object {},
                ],
              },
            },
            Object {
              "name": "merging",
              "value": Object {
                "enum": "SPACES",
                "type": "enum",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций встроенных объектов с неизвестным аргументом 1`] = `"Неизвестный аргумент 'b'"`;

exports[`вызовы функций встроенных объектов с неподходящим типом аргумента 1`] = `"Аргумент с типом текст не может быть установлен параметру с типом число"`;

exports[`вызовы функций переменных обычных 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 5,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_increment",
          "values": Array [
            Object {
              "name": "number",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_decrement",
          "values": Array [
            Object {
              "name": "number",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций переменных с неизвестным аргументом 1`] = `"Неизвестный аргумент 'min2'"`;

exports[`вызовы функций переменных с неподходящим типом аргумента 1`] = `"Аргумент с типом текст не может быть установлен параметру с типом число"`;

exports[`вызовы функций переменных с установкой значения 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "number": 5,
                "type": "number",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_clamp",
          "values": Array [
            Object {
              "name": "min",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
          ],
        },
        Object {
          "action": "set_variable_clamp",
          "values": Array [
            Object {
              "name": "min",
              "value": Object {
                "number": 1,
                "type": "number",
              },
            },
            Object {
              "name": "max",
              "value": Object {
                "number": 2,
                "type": "number",
              },
            },
            Object {
              "name": "number",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "c",
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций фабрик обычных 1`] = `
Object {
  "handlers": Array [
    Object {
      "event": "world_start",
      "operations": Array [
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "a",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "pitch": 0,
                "type": "location",
                "x": 1,
                "y": 2,
                "yaw": 90,
                "z": 3,
              },
            },
          ],
        },
        Object {
          "action": "set_variable_value",
          "values": Array [
            Object {
              "name": "variable",
              "value": Object {
                "scope": "local",
                "type": "variable",
                "variable": "b",
              },
            },
            Object {
              "name": "value",
              "value": Object {
                "pitch": 5,
                "type": "location",
                "x": 1,
                "y": 2,
                "yaw": 4,
                "z": 3,
              },
            },
          ],
        },
      ],
      "position": 0,
      "type": "event",
    },
  ],
}
`;

exports[`вызовы функций фабрик с неподходящим типом аргумента 1`] = `"Аргумент с типом текст не может быть установлен параметру с типом число*"`;
